server/src/main/java/org/elasticsearch/action/search/DfsQueryPhase.java

```diff
@@ -115,4 +127,18 @@ public void onFailure(Exception exception) {
             );
         }
     }
+
+    private void addKnnResultsToQuery(SearchSourceBuilder source) {
+        if (source == null || source.knnSearch() == null) {
+            return;
+        }
+
+        ScoreDocQueryBuilder knnQueryBuilder = new ScoreDocQueryBuilder(knnResults.scoreDocs());
+        if (source.query() == null) {
```

Comment:
In the follow up we can clarify in the documentation that if `query` is not set, it means we are running only knn search.  Because a usual search without a query set implies `match_all` query.